### DISCUSSION TITLE
Periodic BCs with split mesh

### DISCUSSION THREAD

**Discussion by Eilloo** [idaholab/moose#25131] [Q&A General]
Hi all,

Should periodic BCs work with a split mesh?
I have a case where I am loading in a mesh file which is periodic, and translational BCs are accepted as expected when the mesh is replicated, and also if distributed without pre-splitting the mesh.
However, if I pre-split the mesh, I get the following error:
`Periodic boundary neighbor not found`
Is this behaviour expected, and is there any way to apply periodic BCs to large meshes which need splitting?

Thanks!

**Comment by lindsayad** [comment]
I am pretty sure there is a discussion post about this somewhere involving @ttruster. The issue is that when you split the mesh, the ghosting functors for the periodic neighbors are likely not run. If you ran your splitting mesh input with all the algebraic objects from your simulation input, even though you are not actually using them, I am curious whether things would work, since they should add their ghosting needs for consideration when the mesh is split.

**Reply by ttruster** [reply]
Yes the discussion is in here: https://github.com/idaholab/moose/issues/24140. The current framework functions if the mesh is split after the periodic boundaries are loaded, i.e. if it is split at runtime.

**Reply by Eilloo** [reply]
Ah, nice, that makes sense. Is running mesh splitting with all algebraic objects something I can try? I haven't seen anything in the docs, so not sure how you'd go about this...

**Reply by lindsayad** [reply]
Just use the exact same input file as you would use to run your simulation, use the mesh generators you use to generate your mesh, and run that with `--mesh-only`

**Reply by Eilloo** [reply]
Ah I see, I had already tried this - unfortunately no difference, periodic BCs are still not picked up.

**Reply by lindsayad** [reply]
that's very interesting. That's something we should debug. I'm going to transfer this to an issue

**Reply by Eilloo** [reply]
Okay - I have double checked this morning, so just to be extra clear on what does and doesn't work:
- Replicated mesh work fine
- Loading the raw mesh file with `FileMeshGenerator`, defining other subdomains etc in the `Mesh` block, and later specifying a translational periodic BC works fine with `parallel_type = DISTRIBUTED` and `--distributed-mesh`. (Maybe interesting is I can't get the auto-detection to work even though the translation is orthogonal. I wondered if this only works with Moose generated meshes?)
- Running this same file with `--mesh-only`, then using `FileMeshGenerator` to load in the resulting exodus file also works fine with `parallel_type = DISTRIBUTED` and `--distributed-mesh`.
- Using --split-mesh for either of these cases (the raw mesh file, or the --mesh-only exodus file) and loading the resulting .cpr files in `FileMeshGenerator` does not work. Both cases result in the error `Periodic boundary neighbor not found`.

So it seems that it is fine with anything where the mesh is replicated at some stage, even if parts are then deleted on each processor. However, anything where the mesh is pre-split does not work.

I haven't tried this with any other mesh formats yet. These tests were all using the gmsh `.msh` format.

**Reply by lindsayad** [reply]
I've opened https://github.com/idaholab/moose/issues/25158 to track this

**Reply by Eilloo** [reply]
Thanks - thinking about this, there is some other behaviour which could be relevant here, which is when I see the error message:

If I attempt to use auto-detected BCs, I get an error saying they could not be found immediately.

However, specifying the translation, I typically have to wait longer for the error to appear, and other information is output first - for example, `TensorMechanics Action: selecting 'incremental finite strain' formulation` in a simulation with tensor mechanics.

I'm not sure what order everything is supposed to happen, but I wonder if only the automatically detected periodic BCs are handled when other algebraic objects are, but for some reason other types such as translational are not? 
It seems strange that it appears to be searching for the periodic neighbours at different times depending on the formulation. Since the translational seems to happen later than some other initial steps, this could be linked to why it is not implemented when running --mesh-only
